core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -26,42 +26,33 @@
 import org.elasticsearch.indices.query.IndicesQueriesRegistry;
 
 import java.io.IOException;
+import java.util.Objects;
 
 public class QueryParseContext {
 
     private static final ParseField CACHE = new ParseField("_cache").withAllDeprecated("Elasticsearch makes its own caching decisions");
     private static final ParseField CACHE_KEY = new ParseField("_cache_key").withAllDeprecated("Filters are always used as cache keys");
 
-    private XContentParser parser;
-    private ParseFieldMatcher parseFieldMatcher = ParseFieldMatcher.EMPTY;
+    private final XContentParser parser;
 
     private IndicesQueriesRegistry indicesQueriesRegistry;
 
-    public QueryParseContext(IndicesQueriesRegistry registry) {
-        this.indicesQueriesRegistry = registry;
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
+        this.indicesQueriesRegistry = Objects.requireNonNull(registry, "indices queries registry cannot be null");
+        this.parser = Objects.requireNonNull(parser, "parser cannot be null");
+        this.parser.setParseFieldMatcher(Objects.requireNonNull(parseFieldMatcher, "parse field matcher cannot be null"));
```

Comment:
As I said, ideally I'd say its the parser, but thats a huge change. Also one might argue that the parsers job is to keep state for parsing, but the matcher is something the client (e.g. the builders fromXContent) use, so there is just the question of how to transport it there, so that would be the context. I'm just looking into how difficult it would be to remove the parseField from the parser again, this duplication is a source of errors.